### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.FreeModule.Rank

### PR DESCRIPTION
**Pull request by mattrobball**
This PR also fixes the order of universes in `Cardinal.lift_lift`.

---

Feel free to jump in.

* [`linear_algebra.free_module.rank`@`5aa3c1de9f3c642eac76e11071c852766f220fd0`..`465d4301d8da5945ef1dc1b29fb34c2f2b315ac4`](https://leanprover-community.github.io/mathlib-port-status/file/linear_algebra/free_module/rank?range=5aa3c1de9f3c642eac76e11071c852766f220fd0..465d4301d8da5945ef1dc1b29fb34c2f2b315ac4)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
I fixed the build but it's insanely slow right now.

File: Mathlib/LinearAlgebra/FreeModule/Rank.lean [hunk 1]
```diff
@@ -106,23 +109,35 @@ variable [AddCommGroup N] [Module R N] [Module.Free R N]
...
+  have h₁ := @LinearEquiv.lift_rank_eq.{u, max v w, max (max u v) w}
+     (R := R) (M := M ⊗[R] N)
+     (M' := (ChooseBasisIndex R M →₀ R) ⊗[R] (ChooseBasisIndex R N →₀ R)) _ _ _ _
+     (@instModuleTensorProductToSemiringAddCommMonoid _ _ _ _
+       AddCommGroup.toAddCommMonoid
+       AddCommGroup.toAddCommMonoid _ _) fg
+  let _b : Basis (ChooseBasisIndex R M × ChooseBasisIndex R N) R (_ →₀ R) := Finsupp.basisSingleOne
+  rw [LinearEquiv.rank_eq (finsuppTensorFinsupp' R (ChooseBasisIndex R M) (ChooseBasisIndex R N)),
+    ← _b.mk_eq_rank, mk_prod] at h₁
+  rw [lift_inj.1 h₁, rank_eq_card_chooseBasisIndex R M, rank_eq_card_chooseBasisIndex R N]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/FreeModule/Rank.lean]: This should have a much shorter proof in terms of `basis.tensor_product`

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/FreeModule/Rank.lean]: https://github.com/leanprover-community/mathlib/pull/18804

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**